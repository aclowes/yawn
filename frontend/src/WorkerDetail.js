import React from 'react';
import {Panel, PanelGroup, Alert} from 'react-bootstrap';

import API from "./API";
import ExecutionTable from "./ExecutionTable";

export default class WorkerDetail extends React.Component {
  constructor(props) {
    super(props);
    this.state = {worker: null, executions: null, error: null};
  }

  componentDidMount() {
    API.get(`/api/workers/${this.props.params.id}/`, (payload, error) => {
      this.setState({worker: payload, error});

      if (payload) {  // only get if worker returns... not strictly necessary
        document.title = `YAWN - Worker: ${payload.name}`;
        API.get(`/api/executions/?worker=${this.props.params.id}`, (payload, error) => {
          this.setState({executions: payload, error});
        });
      }

    });
  }

  render() {
    const worker = this.state.worker;
    if (worker) {
      return (
        <PanelGroup>
          <Panel header="Details">
            <dl className="dl-horizontal">
              <dt>Name</dt>
              <dd>{worker.name}</dd>
              <dt>Status</dt>
              <dd>{worker.status}</dd>
              <dt>Start Timestamp</dt>
              <dd>{worker.start_timestamp}</dd>
              <dt>Last Heartbeat</dt>
              <dd>{worker.last_heartbeat}</dd>
            </dl>
          </Panel>
          <Panel header="Executions">
            {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
            <ExecutionTable executions={this.state.executions}/>
          </Panel>
        </PanelGroup>
      );
    } else {
      return (
        <div>
          {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
          Loading...
        </div>
      )
    }
  }
}
