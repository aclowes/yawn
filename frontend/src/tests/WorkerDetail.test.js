import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import WorkerDetail from '../WorkerDetail';
import {mockAPI} from './mocks'

it('WorkerDetail loading', () => {
  mockAPI([null]);
  const component = ReactTestRenderer.create(<WorkerDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('WorkerDetail failure', () => {
  mockAPI();
  const component = ReactTestRenderer.create(<WorkerDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('WorkerDetail success', () => {
  mockAPI([
    // the worker
    {
      "id": 4,
      "name": "alec-graybeard.local 36470",
      "status": "exited",
      "start_timestamp": "2017-01-30T04:13:47.966080Z",
      "last_heartbeat": "2017-01-30T04:15:10.845289Z"
    },
    // executions
    [{
      "id": 16,
      "task": {
        "id": 12,
        "name": "done",
        "workflow": {"id": 1, "name": "Simple Workflow Example", "version": 1},
        "status": "succeeded"
      },
      "worker": {
        "id": 7,
        "name": "alec-graybeard.local 36561",
        "status": "exited",
        "start_timestamp": "2017-01-30T04:17:01.039091Z",
        "last_heartbeat": "2017-01-30T04:17:43.763294Z"
      },
      "status": "succeeded",
      "start_timestamp": "2017-01-30T04:17:03.806599Z",
      "minutes_running": null
    }]
  ]);
  const component = ReactTestRenderer.create(<WorkerDetail params={{id: 1}}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
