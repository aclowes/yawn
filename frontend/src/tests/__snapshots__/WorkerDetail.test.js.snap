// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkerDetail failure 1`] = `
<div>
  <div
    className="alert alert-danger"
    role="alert"
  >
    some error message
  </div>
  Loading...
</div>
`;

exports[`WorkerDetail loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`WorkerDetail success 1`] = `
<div
  className="panel-group"
>
  <div
    className="panel panel-default"
    id={undefined}
  >
    <div
      className="panel-heading"
    >
      Details
    </div>
    <div
      className="panel-body"
    >
      <dl
        className="dl-horizontal"
      >
        <dt>
          Name
        </dt>
        <dd>
          alec-graybeard.local 36470
        </dd>
        <dt>
          Status
        </dt>
        <dd>
          exited
        </dd>
        <dt>
          Start Timestamp
        </dt>
        <dd>
          2017-01-30T04:13:47.966080Z
        </dd>
        <dt>
          Last Heartbeat
        </dt>
        <dd>
          2017-01-30T04:15:10.845289Z
        </dd>
      </dl>
    </div>
  </div>
  <div
    className="panel panel-default"
    id={undefined}
  >
    <div
      className="panel-heading"
    >
      Executions
    </div>
    <div
      className="panel-body"
    >
      <table
        className="table table-striped table-bordered table-condensed table-hover"
      >
        <thead>
          <tr>
            <th>
              Worker
            </th>
            <th>
              Workflow
            </th>
            <th>
              Task
            </th>
            <th>
              Status
            </th>
            <th>
              Start Time
            </th>
            <th>
              Runtime
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                onClick={[Function]}
                style={Object {}}
              >
                alec-graybeard.local 36561
              </a>
            </td>
            <td>
              Simple Workflow Example
            </td>
            <td>
              <a
                onClick={[Function]}
                style={Object {}}
              >
                done
              </a>
            </td>
            <td>
              succeeded
            </td>
            <td>
              2017-01-30T04:17:03.806599Z
            </td>
            <td />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
